spec/units/clever_tap_client_spec.rb

```diff
@@ -38,4 +39,272 @@
       end
     end
   end
+
+  describe '#new' do
+    let(:identity_field) { 'ID' }
+    let(:account_id) { 'ABC1234' }
+    let(:account_passcode) { 'AcCPasScoDe123' }
+
+    context 'when credentials set in `CleverTap.setup`' do
+      subject { CleverTap::Client.new }
+
+      before do
+        CleverTap.setup do |config|
+          config.identity_field = identity_field
+          config.account_id = account_id
+          config.account_passcode = account_passcode
+        end
+      end
+
+      it_behaves_like 'configured `Client`'
+    end
+
+    context 'when credentials set in `CleverTap::Client.new`' do
+      subject { CleverTap::Client.new(account_id, account_passcode) }
+      it_behaves_like 'configured `Client`'
+    end
+  end
+
+  def event_factory(id, name)
+    CleverTap::Event.new(
+      data: { 'FBID' => id.to_s, 'Name' => name.to_s },
+      name: 'Web Event',
+      identity: 'FBID'
+    )
+  end
+
+  def profile_factory
+    CleverTap::Profile.new(
+      data: { 'ID' => '1414', 'Name' => 'John', 'Phone' => '+44+441234' }
+    )
+  end
+
+  def setup_clevertap_client
+    CleverTap.setup do |c|
+      c.identity_field = 'ID'
+      c.account_id = AUTH_ACCOUNT_ID
+      c.account_passcode = AUTH_PASSCODE
+    end
+  end
+
+  describe '#batched_upload' do
+    let(:success_proc) { proc { 'sample proc' } }
+
+    subject do
+      client = described_class.new
+      client.on_successful_upload(&success_proc)
+      client.send(:batched_upload, CleverTap::Event, [event1, event2], 0)
+    end
+
+    before { setup_clevertap_client }
+
+    context 'when upload records are homogenous' do
+      context 'and valid records' do
+        let(:event1) { event_factory('1414', 'John') }
+        let(:event2) { event_factory('1515', 'Jill') }
+
+        context 'and objects fit `upload_limit`' do
+          VCR.use_cassette('upload within Event limit') do
+            it_behaves_like 'it calls success hook', 'once'
+            it_behaves_like 'it yields control', 'once'
+          end
+        end
+
+        context 'and objects do not fit `upload_limit`' do
+          before do
+            allow(CleverTap::Event)
+              .to receive(:upload_limit).and_return(1)
+          end
+
+          VCR.use_cassette('upload out of Event limit') do
+            it_behaves_like 'it calls success hook', 'twice'
+            it_behaves_like 'it yields control', 'twice'
+          end
+        end
+      end
+
+      context 'and invalid records' do
+        let(:failure_proc) { proc { 'sample proc' } }
+
+        subject do
+          client = described_class.new
+          client.on_failed_upload(&failure_proc)
+          client.send(:batched_upload, CleverTap::Profile, [profile], 0)
+        end
+
+        let(:profile) { profile_factory }
+
+        it_behaves_like 'it calls failure hook', 'once'
```

Comment:
it's very easy to make tests much much worse by using shared example, so they should be used with extreme care.

Also, think about what you are testing. Don't strive for coverage (unless you can get 100% mutation coverage as well). You need just couple of examples that upon failure the hook is called and than nobody else cares about hooks. Don't repeat testing the same thing over and over again